### ISSUE TITLE
Questions about reproducing FID results of MMDiT+REPA on MSOCCO dataset (Table 12)

### ISSUE THREAD

**Issue by zhuyu-cs** [sihyun-yu/REPA#34]
Hi,

I am trying to reproduce your REPA+MMDiT results on MSOCCO dataset (Table 12), but I cannot achieve the reported baseline FID score of 6.05.

My current setup:
I trained the model for 400k iterations using the default config from your repository (without REPA). For testing, I used an ODE solver with 50 steps on the full COCO validation set. With cfg scale=1.5, I got FID=20.17. Even with scale=2.0, the FID is still much higher than 6.05.

Could you please confirm the exact settings you used to achieve the results in your paper?

Thank you!

**Comment by sihyun-yu** [comment]
Hi, could you try training the model for 150K iteration?

**Comment by zhuyu-cs** [comment]
Thanks for your quick reply! I'll try it right away.

**Comment by zhuyu-cs** [comment]
Thank you for your help. I am currently using the baseline (MMDiT) checkpoint at 150k iterations with cfg 2.0, which gives me an FID of 8.35. I noticed in the UViT codebase (https://github.com/baofff/U-ViT/blob/main/scripts/extract_mscoco_feature.py) that they don't use EMA weights to obtain the image latents (which is the setting I have been following). Could this preprocessing difference be the cause of the gap?

**Comment by zhuyu-cs** [comment]
Could you please help verify if my implementation settings are correct? If there are any misconfigurations, I would greatly appreciate your guidance and corrections.

**Comment by sihyun-yu** [comment]
It's weird. Could you try evaluation by adding --heun option?

**Comment by zhuyu-cs** [comment]
Thank you very much for your help! The two points you mentioned are very important. According to my training log, the proj loss is around -0.33. Additionally, during preprocessing, I didn't resave the images as 256x256, but instead just sorted and renamed the original images and moved them to the directory matching the npy files (since UViT didn't resize and save the original images, I didn't process them either). Then I modified the dataloader:

```python
class MSCOCOFeatureDataset(Dataset):
    def __init__(self, root, image_size=256): 
        self.root = root
        self.num_data, self.n_captions = get_feature_dir_info(root)
        self.image_size = image_size
        
        # newly added 
        self.transform = transforms.Compose([
            transforms.Resize(image_size, interpolation=transforms.InterpolationMode.LANCZOS),  
            transforms.CenterCrop(image_size),  
            transforms.ToTensor()
        ])

    def __len__(self):
        return self.num_data

    def __getitem__(self, index):
        with open(os.path.join(self.root, f'{index}.png'), 'rb') as f:
            img = PIL.Image.open(f).convert('RGB')
            x = self.transform(img)

        z = np.load(os.path.join(self.root, f'{index}.npy'))
        k = random.randint(0, self.n_captions[index] - 1)
        c = np.load(os.path.join(self.root, f'{index}_{k}.npy'))
        
        return x, z, c
```

The preprocessing of the original images might have seriously affected REPA's performance.

**Comment by sihyun-yu** [comment]
I think that is be a problem. The proj_loss should be around -0.7; here is my log for mmdit + repa.
<img width="511" alt="Screenshot 2025-01-03 at 8 59 21 PM" src="https://github.com/user-attachments/assets/5c316922-efa6-4f4f-8bae-a8278cea2c19" />

And I think you can reproduce a similar FID for MMDiT (150K; w/ ema) with --heun option; because it always have shown slightly better results

**Comment by zhuyu-cs** [comment]
Thank you so much for your help. With your reference now, I think I can solve this problem!

**Comment by dlsrbgg33** [comment]
@zhuyu-cs did you use EMA weights of VAE encoder?
wondering if you were able to reproduce the score with your setting

**Comment by zhuyu-cs** [comment]
Yes, I can reproduce the results. Using cfg=2.0, the baseline MMDit achieved FID=6.27, while with REPA added, the FID improved to 4.96. I used non-EMA weights for preprocessing, but followed REPA's settings during decoding by using VAE's EMA weights to recover the images.

ps. I suggest using VAE's EMA weights during preprocessing as well, as this would be more suitable for REPA's scaling factor of 0.18215.

**Comment by dlsrbgg33** [comment]
Thank you, I think it worked after changing the VAE model with EMA version.

**Comment by gstoica27** [comment]
Hi all, 

I was wondering what the data preprocessing procedure should be exactly? I followed the preprocessing instructions described in the README, but the train script doesn't work because of index mismatch. While this is easy to resolve, is this the only thing we need to account for? 

I also see discussion about explicitly saving to 256x256. Is this part of the preprocessing? 

It would be really helpful to understand what the exact procedure is, it's a little confusing at the moment.

Thanks!

**Comment by zhuyu-cs** [comment]
I think you can use https://github.com/baofff/U-ViT/blob/main/scripts/extract_mscoco_feature.py, and just add the following code after line 49 to save the image in the new folder, using the corresponding index: 
```
img = Image.fromarray(img)
img.save(os.path.join(save_dir, f'{idx}.png'))
```

**Comment by gstoica27** [comment]
Ah okay! Did you also resize the images to 256x256 before running this script (and if so was this with the [dataset_tools.py](https://github.com/sihyun-yu/REPA/blob/main/preprocessing/dataset_tools.py) script)? Or did you use the raw images?

**Comment by zhuyu-cs** [comment]
Yes, I have resized the images to 256x256 using this dataloader: https://github.com/baofff/U-ViT/blob/main/datasets.py. Please note that this preprocessing step only depends on the U-ViT codebase.

**Comment by Sierkinhane** [comment]
@zhuyu-cs Hi, zhuyu. May I know how many samples you used for evaluation? I tried 40k samples and got FID around 4.9; 30k samples and got FID around 5.2.

**Comment by zhuyu-cs** [comment]
> [@zhuyu-cs](https://github.com/zhuyu-cs) Hi, zhuyu. May I know how many samples you used for evaluation? I tried 40k samples and got FID around 4.9; 30k samples and got FID around 5.2.

It should be 40k images. https://github.com/sihyun-yu/REPA/blob/67f714503e3892f993844aab088ffc5791c92613/generate_t2i.py#L192

**Comment by Sierkinhane** [comment]
Got it.

**Comment by Sierkinhane** [comment]
> > [@zhuyu-cs](https://github.com/zhuyu-cs) Hi, zhuyu. May I know how many samples you used for evaluation? I tried 40k samples and got FID around 4.9; 30k samples and got FID around 5.2.
> 
> It should be 40k images.
> 
> [REPA/generate_t2i.py](https://github.com/sihyun-yu/REPA/blob/67f714503e3892f993844aab088ffc5791c92613/generate_t2i.py#L192)
> 
> Line 192 in [67f7145](/sihyun-yu/REPA/commit/67f714503e3892f993844aab088ffc5791c92613)
> 
>  create_npz_from_sample_folder(sample_folder_dir, 40192)

Got it. Thank you!
